cpp/src/parquet/annotations-test.cc

```diff
@@ -0,0 +1,1482 @@
+  ASSERT_ANY_THROW(IntAnnotation::Make(-1, false));
+  ASSERT_ANY_THROW(IntAnnotation::Make(0, false));
+  ASSERT_ANY_THROW(IntAnnotation::Make(1, false));
+  ASSERT_ANY_THROW(IntAnnotation::Make(65, false));
+  // Invalid precisions ...
+  ASSERT_ANY_THROW(DecimalAnnotation::Make(-1));
+  ASSERT_ANY_THROW(DecimalAnnotation::Make(0));
+  ASSERT_ANY_THROW(DecimalAnnotation::Make(0, 0));
+  // Invalid scales ...
+  ASSERT_ANY_THROW(DecimalAnnotation::Make(10, -1));
+  ASSERT_ANY_THROW(DecimalAnnotation::Make(10, 11));
+}
+
+static void ConfirmAnnotationProperties(
+    const std::shared_ptr<const LogicalAnnotation>& annotation, bool nested,
+    bool serialized, bool valid) {
+  ASSERT_TRUE(annotation->is_nested() == nested)
+      << annotation->ToString() << " annotation has incorrect nested() property";
+  ASSERT_TRUE(annotation->is_serialized() == serialized)
+      << annotation->ToString() << " annotation has incorrect serialized() property";
+  ASSERT_TRUE(annotation->is_valid() == valid)
+      << annotation->ToString() << " annotation has incorrect valid() property";
+  ASSERT_TRUE(annotation->is_nonnested() != nested)
+      << annotation->ToString() << " annotation has incorrect nonnested() property";
+  ASSERT_TRUE(annotation->is_invalid() != valid)
+      << annotation->ToString() << " annotation has incorrect invalid() property";
+  return;
+}
+
+TEST(TestLogicalAnnotationOperation, AnnotationProperties) {
+  // For each annotation type, ensure that the correct general properties are reported
+
+  ConfirmAnnotationProperties(StringAnnotation::Make(), false, true, true);
+  ConfirmAnnotationProperties(MapAnnotation::Make(), true, true, true);
+  ConfirmAnnotationProperties(ListAnnotation::Make(), true, true, true);
+  ConfirmAnnotationProperties(EnumAnnotation::Make(), false, true, true);
+  ConfirmAnnotationProperties(DecimalAnnotation::Make(16, 6), false, true, true);
+  ConfirmAnnotationProperties(DateAnnotation::Make(), false, true, true);
+  ConfirmAnnotationProperties(
+      TimeAnnotation::Make(true, LogicalAnnotation::TimeUnit::MicroSeconds), false, true,
+      true);
+  ConfirmAnnotationProperties(
+      TimestampAnnotation::Make(true, LogicalAnnotation::TimeUnit::MicroSeconds), false,
+      true, true);
+  ConfirmAnnotationProperties(IntervalAnnotation::Make(), false, true, true);
+  ConfirmAnnotationProperties(IntAnnotation::Make(8, false), false, true, true);
+  ConfirmAnnotationProperties(IntAnnotation::Make(64, true), false, true, true);
+  ConfirmAnnotationProperties(NullAnnotation::Make(), false, true, true);
+  ConfirmAnnotationProperties(JSONAnnotation::Make(), false, true, true);
+  ConfirmAnnotationProperties(BSONAnnotation::Make(), false, true, true);
+  ConfirmAnnotationProperties(UUIDAnnotation::Make(), false, true, true);
+  ConfirmAnnotationProperties(NoAnnotation::Make(), false, false, true);
+  ConfirmAnnotationProperties(LogicalAnnotation::UnknownType(), false, false, false);
```

Comment:
This could be expressed as a loop over a POD struct with values like

```
struct Case {
  ...
};

std::vector<Case> cases = {
  {StringAnnotation::Make(), false, true, true},
  ...
};
```

Reply:
Fixed in commit f1e1b8a8eb8147d9e9f02bed3c6f2c9f5910fd44